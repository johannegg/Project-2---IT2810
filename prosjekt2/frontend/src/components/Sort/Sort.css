.sort-container {
    display: inline-block;
    padding: 10px;
    margin: 10px 0;
    border-radius: 10px;
    box-shadow: 0 4px 8px 0 #00000033;
    width: 200px;
    background-color: white;
}

select {
    padding: 8px 12px;
    border: 1px solid #ccc;
    border-radius: 5px;
    background-color: #fff; 
    font-size: 16px;
    color: var(--theme-dark-grey);
    cursor: pointer;
    outline: none;
    transition: border-color 0.3s;
}

select:hover,
select:focus {
    border-color: #888;
}

option {
    font-size: 16px;
    color: var(--theme-dark-grey);
}

.sort-heading{
    margin-top: 0px;
    margin-bottom: 0px;
    text-align: left; 
    font-size: 1.5em;
	font-weight: 500;
}

.sort-options {
    margin: 10px;
    border: none;
    box-shadow: 0 4px 8px 0 #00000033;
}

.sortHeader {
    display: flex;
    align-items: center;
}

.sortIcon {
    font-size: 20px;
}

@media (max-width: 1050px) {
	.sort-container {
		width: 70%;
	}
}

@media (max-width: 600px) {
    .sort-container {
        width: 300px;
    }
}


@media (max-width: 400px) {
    .sort-container {
        width: 190px;
    }

}

@media (prefers-color-scheme: dark) {
    .sort-container {
        background-color: var(--theme-almost-black);
    }
}