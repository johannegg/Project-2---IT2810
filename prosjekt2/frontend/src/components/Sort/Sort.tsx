import type React from "react";
import { FaSort } from "react-icons/fa";
import { SongData } from "../../utils/types/SongTypes";
import "./Sort.css";

type SortProps = {
    songs: SongData[];
    sortOption: string;
    onSortChange: (newSort: string) => void;
};

const Sort: React.FC<SortProps> = ({ sortOption, onSortChange }) => {

    const handleSortChange = (e: React.ChangeEvent<HTMLSelectElement>) => {
        const newSortOption = e.target.value;
        onSortChange(newSortOption);
    };

    return (
        <div className="sort-container">
            <section className="sortHeader">
				<FaSort className="filterSortIcon" />
                <h2 className="sort-heading">Sort by</h2>
			</section>
            <select className="sort-options" value={sortOption} onChange={handleSortChange}>
                <option value="title_asc">Title A-Z</option>
                <option value="title_desc">Title Z-A</option>
                <option value="artist_asc">Artist A-Z</option>
                <option value="artist_desc">Artist Z-A</option>
                <option value="views_desc">Views</option>
            </select>
        </div>
    );
};

export default Sort;
