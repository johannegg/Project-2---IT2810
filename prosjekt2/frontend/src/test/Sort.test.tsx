import { describe, test, vi, afterEach, expect } from "vitest";
import { render, screen, fireEvent, cleanup } from "@testing-library/react";
import Sort from "../components/Sort/Sort";

describe("Sort", () => {
	const mockOnSortChange = vi.fn(); // Mock function for onSortChange

	// Clean up after each test
	afterEach(() => {
		cleanup();
		vi.resetAllMocks();
	});

	// Test 1: Verifies that Sort renders correctly with default sort option
	test("renders Sort with default sort option", () => {
		render(<Sort songs={[]} sortOption="title_asc" onSortChange={mockOnSortChange} />);

		// Assert that the header and select element are rendered
		expect(screen.getByText("Sort by")).toBeInTheDocument();
		expect(screen.getByRole("combobox")).toHaveValue("title_asc");
	});

	// Test 2: Verifies that the sort options are rendered
	test("renders all sort options", () => {
		render(<Sort songs={[]} sortOption="title_asc" onSortChange={mockOnSortChange} />);
	
		// Assert that all options are present and match the aria-labels
		expect(screen.getByRole("option", { name: "Sort by title A to Z" })).toBeInTheDocument();
		expect(screen.getByRole("option", { name: "Sort by title Z to A" })).toBeInTheDocument();
		expect(screen.getByRole("option", { name: "Sort by artist A to Z" })).toBeInTheDocument();
		expect(screen.getByRole("option", { name: "Sort by artist Z to A" })).toBeInTheDocument();
		expect(screen.getByRole("option", { name: "Sort by views descending" })).toBeInTheDocument();
	});
	

	// Test 3: Verifies that onSortChange is called when a new sort option is selected
	test("calls onSortChange when a new sort option is selected", () => {
		render(<Sort songs={[]} sortOption="title_asc" onSortChange={mockOnSortChange} />);

		// Simulate changing the sort option
		const select = screen.getByRole("combobox");
		fireEvent.change(select, { target: { value: "title_desc" } });

		// Assert that the mock function is called with the new sort option
		expect(mockOnSortChange).toHaveBeenCalledWith("title_desc");
	});

	// Test 4: Verifies that Sort renders correctly with a different sort option
	test("renders Sort with a different sort option", () => {
		render(<Sort songs={[]} sortOption="artist_desc" onSortChange={mockOnSortChange} />);

		// Assert that the selected value is correctly set
		expect(screen.getByRole("combobox")).toHaveValue("artist_desc");
	});

	// Test 5: Snapshot test for Sort with default sort option
	test("matches snapshot with default sort option", () => {
		const { container } = render(
			<Sort songs={[]} sortOption="title_asc" onSortChange={mockOnSortChange} />,
		);

		// Assert that the rendered output matches the snapshot
		expect(container).toMatchSnapshot();
	});

	// Test 6: Snapshot test for Sort with a different sort option
	test("matches snapshot with a different sort option", () => {
		const { container } = render(
			<Sort songs={[]} sortOption="views_desc" onSortChange={mockOnSortChange} />,
		);

		// Assert that the rendered output matches the snapshot
		expect(container).toMatchSnapshot();
	});
});
