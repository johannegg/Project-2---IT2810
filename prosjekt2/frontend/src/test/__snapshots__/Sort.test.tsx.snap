// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`Sort > matches snapshot with a different sort option 1`] = `
<div>
  <div
    class="sort-container"
  >
    <section
      class="sortHeader"
    >
      <svg
        aria-label="Sort icon"
        class="filterSortIcon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 320 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
        />
      </svg>
      <h2
        class="sort-heading"
      >
        Sort by
      </h2>
    </section>
    <select
      aria-label="Sort songs by"
      class="sort-options"
      tabindex="0"
    >
      <option
        aria-label="Sort by title A to Z"
        value="title_asc"
      >
        Title A-Z
      </option>
      <option
        aria-label="Sort by title Z to A"
        value="title_desc"
      >
        Title Z-A
      </option>
      <option
        aria-label="Sort by artist A to Z"
        value="artist_asc"
      >
        Artist A-Z
      </option>
      <option
        aria-label="Sort by artist Z to A"
        value="artist_desc"
      >
        Artist Z-A
      </option>
      <option
        aria-label="Sort by views descending"
        value="views_desc"
      >
        Views
      </option>
    </select>
  </div>
</div>
`;

exports[`Sort > matches snapshot with default sort option 1`] = `
<div>
  <div
    class="sort-container"
  >
    <section
      class="sortHeader"
    >
      <svg
        aria-label="Sort icon"
        class="filterSortIcon"
        fill="currentColor"
        height="1em"
        stroke="currentColor"
        stroke-width="0"
        viewBox="0 0 320 512"
        width="1em"
        xmlns="http://www.w3.org/2000/svg"
      >
        <path
          d="M41 288h238c21.4 0 32.1 25.9 17 41L177 448c-9.4 9.4-24.6 9.4-33.9 0L24 329c-15.1-15.1-4.4-41 17-41zm255-105L177 64c-9.4-9.4-24.6-9.4-33.9 0L24 183c-15.1 15.1-4.4 41 17 41h238c21.4 0 32.1-25.9 17-41z"
        />
      </svg>
      <h2
        class="sort-heading"
      >
        Sort by
      </h2>
    </section>
    <select
      aria-label="Sort songs by"
      class="sort-options"
      tabindex="0"
    >
      <option
        aria-label="Sort by title A to Z"
        value="title_asc"
      >
        Title A-Z
      </option>
      <option
        aria-label="Sort by title Z to A"
        value="title_desc"
      >
        Title Z-A
      </option>
      <option
        aria-label="Sort by artist A to Z"
        value="artist_asc"
      >
        Artist A-Z
      </option>
      <option
        aria-label="Sort by artist Z to A"
        value="artist_desc"
      >
        Artist Z-A
      </option>
      <option
        aria-label="Sort by views descending"
        value="views_desc"
      >
        Views
      </option>
    </select>
  </div>
</div>
`;
